tests/unit/recommenders/models/test_ncf_dataset.py

```diff
@@ -1,120 +1,190 @@
-        user, item, labels = batch
-        assert len(user) == 1
-        assert len(item) == 1
-        assert len(labels) == 1
-
-        # right labels
-        for u, i, is_pos in zip(user, item, labels):
-            if is_pos:
-                assert i in positive_pool[u]
-            else:
-                assert i not in positive_pool[u]
-
-            label_list.append(is_pos)
-
-    # neagtive smapling
-    assert len(label_list) == (N_NEG + 1) * sum(label_list)
-
-
-def test_test_loader(python_dataset_ncf):
-    train, test = python_dataset_ncf
-    data = Dataset(train=train, test=test, n_neg=N_NEG, n_neg_test=N_NEG_TEST)
-
-    # positive user-item dict, noting that the pool is train+test
-    positive_pool = {}
-    df = train.append(test)
-    for u in df[DEFAULT_USER_COL].unique():
-        positive_pool[u] = set(df[df[DEFAULT_USER_COL] == u][DEFAULT_ITEM_COL])
-
-    for batch in data.test_loader():
-        user, item, labels = batch
-        # shape
-        assert len(user) == N_NEG_TEST + 1
-        assert len(item) == N_NEG_TEST + 1
-        assert len(labels) == N_NEG_TEST + 1
-
-        label_list = []
-
-        for u, i, is_pos in zip(user, item, labels):
-            if is_pos:
-                assert i in positive_pool[u]
-            else:
-                assert i not in positive_pool[u]
-
-            label_list.append(is_pos)
-
-        # leave-one-out
-        assert sum(label_list) == 1
-        # right labels
-        assert len(label_list) == (N_NEG_TEST + 1) * sum(label_list)
+    assert set(datafile.users) == set(users)
+    assert set(datafile.items) == set(items)
+    assert set(datafile.user2id.keys()) == set(users)
+    assert set(datafile.item2id.keys()) == set(items)
+    assert len(set(datafile.user2id.values())) == len(users)
+    assert len(set(datafile.item2id.values())) == len(items)
+    assert datafile.data_len == train.shape[0]
+
+    datafile_records = []
+    with datafile as f:
+        for line in f:
+            datafile_records.append({
+                DEFAULT_USER_COL: line[DEFAULT_USER_COL],
+                DEFAULT_ITEM_COL: line[DEFAULT_ITEM_COL],
+                DEFAULT_RATING_COL: line[DEFAULT_RATING_COL]
+            })
+    datafile_df = pd.DataFrame.from_records(datafile_records)
+    assert datafile_df.shape[0] == train.shape[0]
+
+    # test the data loaded from the file is the same as original data
+    datafile_df = datafile_df.sort_values(by=[DEFAULT_USER_COL, DEFAULT_ITEM_COL])
+    train = train.sort_values(by=[DEFAULT_USER_COL, DEFAULT_ITEM_COL])
+    train[DEFAULT_RATING_COL] = train[DEFAULT_RATING_COL].apply(lambda x: float(x > 0))
+    train = train.drop(DEFAULT_TIMESTAMP_COL, axis=1)
+    assert train.equals(datafile_df)
+
+    # test data can be loaded for a valid user and it throws exception for invalid user
+    user = train[DEFAULT_USER_COL].iloc[0]
+    missing_user = train[DEFAULT_USER_COL].iloc[0] + 1
```

Comment:
Why is this user missing? Isn't it inside the valid range of user ids (1 - 100)?

Reply:
I'll check this.